### ISSUE TITLE
Bug: [no-confusing-void-expression] should check awaited type of returned value in async function

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10014]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&showAST=types&fileType=.tsx&code=IYZwngdgxgBAZgV2gFwJYHsIwA4Cd0C2qIApgGrqoAmAKugILjQAUAlAFwwAK%2BRpAPADdKVAHwwA3gCgYMXCWQJcWHoWIkAdPJDoANoJJsA3FIC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90UAO4BDaE0jgwAXxD6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
async function promiseVoidToAsync(): Promise<void> {
  return Promise.resolve();
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-confusing-void-expression": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect that the return statement should error for returning `void`

### Actual Result

No error

### Additional Info

Note that this is (correctly) flagged:
```ts
async function promiseVoidToAsync(): Promise<void> {
  return await Promise.resolve();
}
```

**Comment by bradzacher** [comment]
```ts
async function promiseVoidToAsync(): Promise<void> {
  return Promise.resolve();
}
```

Why is this code bad? The return type of the expression is `Promise<void>` -- so you're returning a value that exactly matches the return type and it is non-void.

```ts
async function promiseVoidToAsync(): Promise<void> {
  return await Promise.resolve();
}
```

This makes sense as an error because the expression returns `void`, not `Promise<void>`

**Comment by kirkwaiblinger** [comment]
in short... I figure these things are basically the same due to the `async` semantics, so they should be flagged the same (see https://github.com/typescript-eslint/typescript-eslint/issues/8674 for the same principle at work in another context)

As for a more a pragmatic justification, I think it's better and more within the spirit of the rule to always do
```ts
async function PromiseVoidToAsync(): Promise:void> {
    await Promise.resolve();
}
```

**Comment by bradzacher** [comment]
They're not the same though yeah. The returned values resolves to the same thing - but they're semantically different.

`return Promise` will return the promise and not await it in the context of the current function.
`return await Promise` will await the promise in the context of the current function and will then return the value to the parent context.

I think that it is more correct to await the promise and have no return - but I don't think such a stylistic decision should be enforced by this rule - that's why we have `return-await`.

**Comment by kirkwaiblinger** [comment]
Fair enough 🤷‍♂️
